src/data/file_preprocessing/echo_files_parser.py

```diff
@@ -0,0 +1,115 @@
+        """
+        self.echo_files = echo_files
+        self.echo_df = pd.DataFrame()
+        self.exceptions_df = pd.DataFrame()
+
+    def find_marker_rows(self, file:str, markers:tuple[str]) -> list[int]:
+        """
+        Finds the row numbers of marker lines in a file.
+
+        :param file: file to search
+        :param markers: markers to search for
+        """
+        with open(file) as f:
+            markers_rows = list()
+            for i, line in enumerate(f):
+                if line.strip() in markers:
+                    markers_rows.append(i)
+                if len(markers_rows) == len(markers):
+                    return markers_rows
+        if len(markers_rows) == 0:
+            raise ValueError('No marker found in file.')
+        return markers_rows
+
+    def parse_files(self) -> EchoFilesParser:
+        """
+        Preprocesses csv echo files, splits regular records from exceptions.
+        """
+        if not(all(file.endswith('.csv') for file in self.echo_files)):
+            raise ValueError(f"Expected files to be of '*.csv' type, provided: {self.echo_files}")
+
+        for filename in self.echo_files: 
+            markers = self.find_marker_rows(filename, ('[EXCEPTIONS]', '[DETAILS]'))
+
+            if len(markers) == 2:
+                exceptions_line, details_line = markers
+                exceptions_df = pd.read_csv(filename, skiprows=exceptions_line+1, nrows=(details_line-1)-exceptions_line-2)
+                echo_df = pd.read_csv(filename, skiprows=details_line+1)
+            else:
+                exceptions_df = pd.DataFrame()
+                echo_df = pd.read_csv(filename, skiprows=markers[0]+1)
+
+            echo_df = echo_df[~echo_df[echo_df.columns[0]].str.lower().str.startswith('instrument')]
+
+            self.exceptions_df = exceptions_df if self.exceptions_df.empty else pd.concat([self.exceptions_df, exceptions_df], ignore_index=True)
+            self.echo_df = echo_df if self.echo_df.empty else pd.concat([self.echo_df, echo_df], ignore_index=True)
+
+        return self
+    
+    def link_bmg_files(self, 
+                       bmg_files: list[str],
+                       bmg_columns: list[str] = ['Well', 'Value'],
+                       bmg_keys: list[str] = ['Plate_barcode', 'Well'],
+                       echo_keys: list[str] = ['Destination Plate Barcode','Destination Well']) -> EchoFilesParser:
+        """
+        Links bmg files to the echo files.
+
+        :param bmg_files: list of bmg files
+        :param echo_keys: list of echo keys to merge on
+        :param bmg_keys: list of bmg keys to merge on       
+        """
+        echo_bmg_linked_df = pd.DataFrame()
+
+        for bmg_file in bmg_files:
+            plate_barcode = bmg_file.split('.')[0].split('\\')[-1]
+            bmg_df = pd.read_csv(bmg_file, sep='\t', names=bmg_columns)
+            bmg_df[bmg_keys[0]]=plate_barcode
+            echo_bmg_linked_df = self.echo_df.merge(bmg_df, 
```

Comment:
This line is quite a bit hard to read, imo it'd be nice to use the usuall if statement rather than a ternary operator

Reply:
Agree, changed the approach though (appending to the list as in `echo_dfs`)